### PR TITLE
Dev utest

### PR DESCRIPTION
**Pull request by MirrorIce**
This is the first unit test made for the CollectionReader and StorageWriter

### CODE REVIEW

**Comment by bbferka** [comment]
@MirrorIce  remove all autogenerated files, and keep the PR as simple as possible. (e.g. revert all changes to files that are not related to the PR; Also: name the PRs such that anyone can easily understand what is the contribution: i.e. in your case it should be something like: added unit test for reading and writing images to mongo;

**Comment by bbferka** [comment]
@MirrorIce  remove all autogenerated files, and keep the PR as simple as possible. (e.g. revert all changes to files that are not related to the PR; Also: name the PRs such that anyone can easily understand what is the contribution: i.e. in your case it should be something like: added unit test for reading and writing images to mongo;

File: descriptors/analysis_engines/demo.xml [hunk 1]
```diff
@@ -174,14 +174,6 @@
...
        <node>CollectionReader</node>
-       <node>ImagePreprocessor</node>
-       <node>PointCloudFilter</node>
-       <node>NormalEstimator</node>
-       <node>PlaneAnnotator</node>
-       <node>ImageSegmentationAnnotator</node>
-       <node>PointCloudClusterExtractor</node>
-       <node>ClusterMerger</node>
-       <node>Cluster3DGeometryAnnotator</node>	
        <node>StorageWriter</node>
```
Comment (bbferka) [inline descriptors/analysis_engines/demo.xml]: these changes break the tutorials online; consider using the same AE definitions from the other uTest;

File: config/config_mongodb_continual.ini [hunk 2]
```diff
@@ -3,7 +3,7 @@ interface=MongoDB
 
 [mongodb]
 host=localhost
-db=Scenes
+db=UTest
```
Comment (bbferka) [inline config/config_mongodb_continual.ini]: this might break the tutorials online

File: src/io/CMakeLists.txt [hunk 3]
```diff
@@ -31,3 +31,5 @@ target_link_libraries(rs_CollectionReader rs_core rs_io)
 
 rs_add_library(rs_TFBroadcaster src/TFBroadcaster.cpp)
 target_link_libraries(rs_TFBroadcaster rs_core)
+
+rs_add_executable(automated_u_test src/automated_u_test.cpp)
```
Comment (bbferka) [inline src/io/CMakeLists.txt]: we should not need this right? This is now incorporated into a unit test; No use having it anymore;

File: test/test_reconfig/test_reconfig.cpp [hunk 4]
```diff
@@ -0,0 +1,208 @@
+/**
```
Comment (bbferka) [inline test/test_reconfig/test_reconfig.cpp]: maybe we should leave this file where it was originally, now that you have it in a TEST

File: test/robosherlock_utest/test_utest.cpp [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+    const std::string arg = argv[argI];
+    cout << "This is an argument " << argv[argI] << '\n';
+
+    if(arg == "-?" || arg == "-h" || arg == "--help")
+    {
+      help();
+      return 0;
+    }
+  }*/
+  priv_nh.param("ae", analysisEnginesArg, std::string("demo"));
```
Comment (bbferka) [inline test/robosherlock_utest/test_utest.cpp]: don't use demo; use u_test;

File: test/robosherlock_utest/test_utest.cpp [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+  priv_nh.param("analysis_engines", analysisEnginesArg, analysisEnginesArg);
```
Comment (bbferka) [inline test/robosherlock_utest/test_utest.cpp]: is this needed? I don't think so;

File: test/robosherlock_utest/test_utest.cpp [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+ * limitations under the License.
+ */
+
+
+/**
+ * Error output if program is called with wrong parameter.
+ */
+
+
+void help()
```
Comment (bbferka) [inline test/robosherlock_utest/test_utest.cpp]: no help function is needed; User does not have options on how to run a test;

File: test/robosherlock_utest/test_utest.cpp [hunk 8]
```diff
@@ -0,0 +1,271 @@
+#include <string>
```
Comment (bbferka) [inline test/robosherlock_utest/test_utest.cpp]: I think this should be called differently; test_storage for example

File: test/main.cpp [hunk 9]
```diff
@@ -3,12 +3,27 @@
...
+#include <pcl/point_types.h>
+#include <ros/ros.h>
 
+#include <mongo/client/dbclient.h>
 
+//Variables to be tested
 
-int main(int argc, char **argv)
+		
+int main2(int argc, char **argv)
```
Comment (bbferka) [inline test/main.cpp]: tis looks weird? why main2?

File: src/io/src/CollectionReader.cpp [hunk 10]
```diff
@@ -149,6 +150,7 @@ class CollectionReader : public Annotator
       else if(interface_ == "Camera")
       {
         cameras_.push_back(new ROSCameraBridge(pt));
+	outInfo("Interface is Camera");
```
Comment (bbferka) [inline src/io/src/CollectionReader.cpp]: remove these outputs; you only had them in there for debug purposes;

File: src/core/include/rs/scene_cas.h [hunk 11]
```diff
@@ -38,8 +38,9 @@
 #include <tf/LinearMath/Transform.h>
 
 #include <rs/types/all_types.h>
-#include <rs/conversion/conversion.h>
 #include <rs/view_names.h>
+#include <rs/conversion/conversion.h>
+
```
Comment (bbferka) [inline src/core/include/rs/scene_cas.h]: revert:)

File: src/core/include/rs/conversion/conversion.h [hunk 12]
```diff
@@ -32,6 +32,8 @@
 // RS
 #include <rs/feature_structure_proxy.h>
 
+
+
```
Comment (bbferka) [inline src/core/include/rs/conversion/conversion.h]: revert

File: test/main.cpp (hunk 9 above)
Comment (MirrorIce) [reply test/main.cpp]: Reviewed for the unit test purposes, we used the main method into the unit test

File: src/io/CMakeLists.txt (hunk 3 above)
Comment (MirrorIce) [reply src/io/CMakeLists.txt]: Right, this is no longer needed

File: config/config_mongodb_continual.ini (hunk 2 above)
Comment (MirrorIce) [reply config/config_mongodb_continual.ini]: Should we use a diiferent config file or use the Scenes database name?

Comment (bbferka) [reply config/config_mongodb_continual.ini]: different config files (e.g. data_loader_uTest);  I am preparing a new set of images for this; Fee free to modify that config file to suite your needs.

File: src/core/include/rs/scene_cas.h (hunk 11 above)
Comment (MirrorIce) [reply src/core/include/rs/scene_cas.h]: If I revert it outputs an error: rs::conversion has not been decalared in scene_cas.h

Comment (bbferka) [reply src/core/include/rs/scene_cas.h]: iinteresting...at compile time?

File: CMakeCache.txt [hunk 13]
```diff
@@ -0,0 +1,512 @@
+# This is the CMakeCache file.
+# For build in directory: /home/mirrorice/catkin_ws2/src/robosherlock
```
Comment (bbferka) [inline CMakeCache.txt]: @MirrorIce you forgot this file;  and the one above...they need to go;

File: CMakeLists.txt.user [hunk 14]
```diff
@@ -0,0 +1,766 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bbferka) [inline CMakeLists.txt.user]: This is the Qt Creater config file; remove

File: src/core/include/rs/scene_cas.h (hunk 11 above)
Comment (bbferka) [reply src/core/include/rs/scene_cas.h]: so the order of the inlude matter for you? I just meant that it is not woth adding a change that has no benefit (.e.g. moving the inlcude two lines down)

File: CMakeLists.txt.user.615d87e [hunk 15]
```diff
@@ -0,0 +1,488 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bbferka) [inline CMakeLists.txt.user.615d87e]: delete :)

File: config/config_mongodb_playback.ini [hunk 16]
```diff
@@ -5,7 +5,7 @@ interface=MongoDB
 host=localhost
 db=Scenes
 continual=false
-loop=true
+loop=false
```
Comment (bbferka) [inline config/config_mongodb_playback.ini]: revert

File: descriptors/analysis_engines/demo.xml [hunk 17]
```diff
@@ -145,27 +145,36 @@
           <parameter>CollectionReader/camera_config_files</parameter>
```
Comment (bbferka) [inline descriptors/analysis_engines/demo.xml]: are you still using demo in the u_test? this breaks the tutorials;

File: descriptors/annotators/io/CollectionReader.xml [hunk 18]
```diff
@@ -24,7 +24,8 @@
         <value>
           <array>
 <!--            <string>config_kinect2.ini</string>-->
-            <string>config_kinect_robot.ini</string>
+<!--             <string>config_kinect_robot.ini</string> -->
+		<string>config_mongodb_playback.ini</string>
```
Comment (bbferka) [inline descriptors/annotators/io/CollectionReader.xml]: this alos breaks tutorials

File: descriptors/annotators/io/StorageWriter.xml [hunk 19]
```diff
@@ -50,14 +50,14 @@
       <nameValuePair>
         <name>storagedb</name>
         <value>
-          <string>Scenes</string>
+          <string>UTest</string>
```
Comment (bbferka) [inline descriptors/annotators/io/StorageWriter.xml]: Same here; tutorial break;

File: src/io/CMakeLists.txt (hunk 3 above)
Comment (bbferka) [reply src/io/CMakeLists.txt]: remove file and entry; this is now part of the unit test;

File: test/main.cpp [hunk 20]
```diff
@@ -12,3 +12,14 @@ int main(int argc, char **argv)
   testing::InitGoogleTest(&argc, argv);
   return RUN_ALL_TESTS();
 }
+
```
Comment (bbferka) [inline test/main.cpp]: waddduuup with all the new lines here?

File: test/robosherlock_utest/test_storage.cpp [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+  }
+  catch(const rs::FrameFilterException &){}
+  engine->collectionProcessComplete();
+  cout << "This is the end of the program\n";
+  cout << "Conclusion db image and initial image equality is: " << (db_image_width == initial_image_width);
+  engine->destroy();
+  testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+
+}
```
Comment (bbferka) [inline test/robosherlock_utest/test_storage.cpp]: this looks good so far; remove the help function; it is not needed; Also remove all the extra outputs; they can be extra checks in the Unit test instead;

**Comment by bbferka** [comment]
travis fails on the unti tests (it hangs forever until it crashes)...hopefully #127 will fix this;

File: descriptors/annotators/io/StorageWriter.xml [hunk 22]
```diff
@@ -57,7 +57,7 @@
         <name>enableViews</name>
         <value>
           <array>
-            <string>color_image_hd</string>
+            <string>color_image</string>
```
Comment (bbferka) [inline descriptors/annotators/io/StorageWriter.xml]: revert pls; if you need to change this param overwrite it in the u_test AEl

File: src/io/src/automated_u_test.cpp [hunk 23]
```diff
@@ -0,0 +1,261 @@
+/**
```
Comment (bbferka) [inline src/io/src/automated_u_test.cpp]: this file is not needed; delete;

File: test/main.cpp [hunk 24]
```diff
@@ -4,11 +4,83 @@
...
+  {
+    ros::init(argc, argv, std::string("RoboSherlock"));
+
+  }
+  ros::NodeHandle priv_nh;
+  priv_nh = ros::NodeHandle("~");
+
+  //ros::NodeHandle priv_nh = ros::NodeHandle("~");
+
+  priv_nh.param("ae", analysisEnginesArg, std::string("u_test"));
```
Comment (bbferka) [inline test/main.cpp]: this is not a rosparam; you can set the value of analysisEnginesArg where it is declared (line 17)

File: test/main.h [hunk 25]
```diff
@@ -0,0 +1,13 @@
+#ifndef MAIN_H
+#define MAIN_H
+ /* extern std::string analysisEnginesArg, savePath;
+  extern std::vector<std::string> analysisEngines;
+  extern ros::NodeHandle priv_nh;
+  extern std::vector<std::string> analysisEngineFiles;
+  extern std::ostringstream engineList;
+  extern std::string engine_file;
+  extern uima::ErrorInfo errorInfo; */
+  extern uima::AnalysisEngine *engine;
```
Comment (bbferka) [inline test/main.h]: yeah, as you said; this is a hack:)...we can live with it for now.

File: test/main.cpp [hunk 26]
```diff
@@ -4,11 +4,83 @@
...
+  priv_nh.param("analysis_engines", analysisEnginesArg, analysisEnginesArg);
+  
+  analysisEngines.push_back(analysisEnginesArg);
```
Comment (bbferka) [inline test/main.cpp]: this whole section can be removed; We don't have multple engines; rs::common::getAEPaths(analysisEnginesArg, engine_file);

File: test/robosherlock_utest/test_storage.cpp [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+  
+  
+  
+
+
+  
+  
+
+
+
```
Comment (bbferka) [inline test/robosherlock_utest/test_storage.cpp]: what's with all the new lines?

**Pull request merged**